sdk/core/Azure.Core/src/Shared/ClientDiagnostics.cs

```diff
@@ -45,81 +43,58 @@ internal static HttpMessageSanitizer CreateMessageSanitizer(DiagnosticsOptions d
-        /// <param name="message">The error message.</param>
-        /// <param name="errorCode">The error code.</param>
         /// <param name="additionalInfo">Additional error details.</param>
-        protected virtual void ExtractFailureContent(
+        protected virtual ResponseError? ExtractFailureContent(
             string? content,
             ResponseHeaders responseHeaders,
-            ref string? message,
-            ref string? errorCode,
             ref IDictionary<string, string>? additionalInfo)
         {
-            ExtractAzureErrorContent(content, ref message, ref errorCode);
+            return ExtractAzureErrorContent(content);
         }
 
-        internal static void ExtractAzureErrorContent(
-            string? content,
-            ref string? message,
-            ref string? errorCode)
+        internal static ResponseError? ExtractAzureErrorContent(string? content)
         {
             try
             {
                 // Optimistic check for JSON object we expect
                 if (content == null ||
-                    !content.StartsWith("{", StringComparison.OrdinalIgnoreCase)) return;
-
-                string? parsedMessage = null;
-                using JsonDocument document = JsonDocument.Parse(content);
-                if (document.RootElement.TryGetProperty("error", out var errorProperty))
-                {
-                    if (errorProperty.TryGetProperty("code", out var codeProperty))
-                    {
-                        errorCode = codeProperty.GetString();
-                    }
-                    if (errorProperty.TryGetProperty("message", out var messageProperty))
-                    {
-                        parsedMessage = messageProperty.GetString();
-                    }
-                }
+                    !content.StartsWith("{", StringComparison.OrdinalIgnoreCase)) return null;
 
-                // Make sure we parsed a message so we don't overwrite the value with null
-                if (parsedMessage != null)
-                {
-                    message = parsedMessage;
-                }
+                return JsonSerializer.Deserialize<ErrorResponse>(content)?.Error;
             }
             catch (Exception)
             {
                 // Ignore any failures - unexpected content will be
                 // included verbatim in the detailed error message
             }
+
+            return null;
         }
 
-        public async ValueTask<RequestFailedException> CreateRequestFailedExceptionAsync(Response response, string? message = null, string? errorCode = null, IDictionary<string, string>? additionalInfo = null, Exception? innerException = null)
+        public async ValueTask<RequestFailedException> CreateRequestFailedExceptionAsync(Response response, ResponseError? error = null, IDictionary<string, string>? additionalInfo = null, Exception? innerException = null)
```

Comment:
Should we consider adding an overload that takes `HttpMessage` so that consumers don't need to construct a ResponseError just to call?

Reply:
Is overload is specifically for consumers that want to override the error parsing and pass their own error object.

Reply:
I saw some client changes where they are just constructing a default `ResponseError`